File: bcs-cc/confcenter/apis/serializers.go
/*
 * Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community Edition) available.
 * Copyright (C) 2017-2019 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package apis

import (
	"bcs_cc/storage/models"
	"bcs_cc/utils"
	"fmt"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/gin-gonic/gin/binding"
	"gopkg.in/go-playground/validator.v8"
)

// Pagination :
/* Pagination : pagination start */
type Pagination struct {
	Limit  int `form:"limit"`
	Offset int `form:"offset"`
}

func (p *Pagination) setDefaultLimit() {
	if p.Limit == 0 {
		p.Limit = 20
	}
}

/* pagination end */

/* areas start */
// create area params
type createAreaJSON struct {
	Name          string `binding:"required" json:"name"`
	Description   string `json:"description"`
	Configuration string `binding:"required" json:"configuration"`
	ChineseName   string `binding:"required" json:"chinese_name"`
}

// update area params
type updateAreaJSON struct {
	models.Model
	Description   string `json:"description"`
	Configuration string `json:"configuration"`
	ChineseName   string `json:"chinese_name"`
}

/* areas end */

/* project start*/
// fetch project params
type queryProjectParams struct {
	ProjectIDs     string `form:"project_ids"`
	EnglishNames   string `form:"english_names"`
	ApprovalStatus string `form:"approval_status"`
	Creator        string `form:"creator"`
	ProjectType    string `form:"project_type"`
	IsSecrecy      string `form:"is_secrecy"`
	IsOfflined     string `form:"is_offline"`
	DesireAllData  string `form:"desire_all_data"`
}

type queryprojectData struct {
	ProjectIDList   []string `json:"project_ids"`
	EnglishNameList []string `json:"project_codes"`
}

// create project data
type createProjectData struct {
	ProjectID      string `binding:"required" json:"project_id"`
	Name           string `binding:"required,max=64" json:"project_name"`
	EnglishName    string `binding:"required,max=32" json:"english_name"`
	ProjectType    uint   `binding:"exists" json:"project_type"`
	UseBK          bool   `binding:"exists" json:"use_bk"`
	BGID           uint   `json:"bg_id"`
	BGName         string `json:"bg_name"`
	DeptID         uint   `json:"dept_id"`
	DeptName       string `json:"dept_name"`
	CenterID       uint   `json:"center_id"`
	CenterName     string `json:"center_name"`
	CCAppID        uint   `json:"cc_app_id"`
	Creator        string `binding:"required" json:"creator"`
	Description    string `json:"description"`
	Kind           uint   `json:"kind"`
	DeployTypeList []uint `json:"deploy_type"`
	DataID         uint   `json:"data_id"`
	IsSecrecy      bool   `json:"is_secrecy"`
	LogoAddr       string `json:"logo_addr"`
}

// update project data
type updateProjectData struct {
	Name           string  `binding:"max=64" json:"project_name"`
	Description    string  `json:"description"`
	IsOfflined     *bool   `json:"is_offlined"`
	ProjectType    *uint   `json:"project_type"`
	UseBK          *bool   `json:"use_bk"`
	BGID           *uint   `json:"bg_id"`
	BGName         *string `json:"bg_name"`
	DeptID         *uint   `json:"dept_id"`
	DeptName       *string `json:"dept_name"`
	CenterID       *uint   `json:"center_id"`
	CenterName     *string `json:"center_name"`
	CCAppID        *uint   `json:"cc_app_id"`
	Kind           *uint   `json:"kind"`
	DeployTypeList []uint  `json:"deploy_type"`
	Updator        string  `binding:"required" json:"updator"`
	DataID         *uint   `json:"data_id"`
	IsSecrecy      *bool   `json:"is_secrecy"`
	ApprovalStatus *uint   `json:"approval_status"`
	LogoAddr       *string `json:"logo_addr"`
	Remark         *string `json:"remark"`
	Extra          *string `json:"extra"`
}

/* project end */

/* cluster version start */
type clusterVersionParams struct {
	VerID       string `form:"ver_id"`
	Environment string `form:"environment"`
	Kind        string `form:"kind"`
}

// set the default value for environment
func (params *clusterVersionParams) setVersionEnvDefaultValue(defaultValue string) {
	if params.Environment == "" {
		params.Environment = defaultValue
	}
}

// set the default value for kind
func (params *clusterVersionParams) setVersionKindDefaultValue(defaultValue string) {
	if params.Kind == "" {
		params.Kind = defaultValue
	}
}

// create base version data
type createClusterBaseVersionData struct {
	Creator     string `json:"creator" binding:"required"`
	Kind        string `json:"kind" binding:"required"`
	Version     string `json:"version" binding:"required"`
	SubVersion  string `json:"sub_version"`
	Environment string `json:"environment" binding:"required"`
	Configure   string `json:"configure" binding:"required"`
}

// update base version data
type updateClusterBaseVersionData struct {
	Version     string `json:"version" binding:"required"`
	Environment string `json:"environment" binding:"required"`
	Kind        string `json:"kind" binding:"required"`
	Configure   string `json:"configure" binding:"required"`
}

type createClusterSnapshotData struct {
	Creator   string `json:"creator" binding:"required"`
	ClusterID string `json:"cluster_id" binding:"required"`
	Snapshot  string `json:"configure" binding:"required"`
}

/* cluster version end */

/* master start */
type createMasterData struct {
	InnerIP      string `binding:"required" json:"inner_ip"`
	ExtendedInfo string `json:"extended_info"`
	Backup       string `json:"backup"`
	Hostname     string `json:"hostname"`
	Status       string `json:"status"`
}

// update master info
type updateMasterData struct {
	InnerIP    string `binding:"required" json:"inner_ip"`
	Status     string `binding:"required" json:"status"`
	InstanceID string `json:"instance_id"`
}

type batchCreateMasterData struct {
	NewInnerIPList []createMasterData `binding:"required,dive" json:"new_inner_ip_list"`
}

type replaceMasterParams struct {
	ReplaceInnerIPList []string           `binding:"required" json:"replace_inner_ip_list"`
	NewIPListInfo      []createMasterData `binding:"required,dive" json:"new_inner_ip_list"`
}

/* master end */

/* cluster start */
type clusterListParamsForm struct {
	InnerIP       string `form:"ip"`
	Name          string `form:"name"`
	DesireAllData string `form:"desire_all_data"`
	Pagination
}

type clusterListDataJSON struct {
	ClusterIDList []string `binding:"required" json:"cluster_ids"`
}

type createClusterDataJSON struct {
	ProjectID       string             `json:"project_id"`
	ClusterID       string             `json:"cluster_id"`
	ClusterNum      int                `json:"cluster_num"`
	Name            string             `binding:"required" json:"name"`
	Creator         string             `binding:"required" json:"creator"`
	Description     string             `json:"description"`
	Type            string             `json:"type"`
	Environment     string             `binding:"required" json:"environment"`
	AreaID          int                `binding:"required" json:"area_id"`
	Status          string             `json:"status"`
	ConfigSvrCount  int                `json:"config_svr_count"`
	MasterCount     int                `json:"master_count"`
	Artifactory     string             `json:"artifactory"`
	IPResourceTotal int                `json:"ip_resource_total"`
	IPResourceUsed  int                `json:"ip_resource_used"`
	MasterIPs       []createMasterData `binding:"required,dive" json:"master_ips"`
	NeedNAT         bool               `json:"need_nat"`
	NotNeedNAT      bool               `json:"not_need_nat"`
	ExtraClusterID  string             `json:"extra_cluster_id"`
}

/* cluster end*/

/* node start */
type nodeListParamsForm struct {
	InnerIPS      string `form:"inner_ip_in"`
	DesireAllData string `form:"desire_all_data"`
	Pagination
}

type nodeCreateDataJSON struct {
	Name        string `json:"name"`
	Creator     string `binding:"required" json:"creator"`
	Description string `json:"description"`
	InnerIP     string `binding:"required" json:"inner_ip"`
	OutterIP    string `json:"outter_ip"`
	Kind        string `json:"kind"`
	DeviceClass string `json:"device_class"`
	InstanceID  string `json:"instance_id"`
}

type nodeUpdateDataJSON struct {
	InnerIP     string `binding:"required" json:"inner_ip"`
	Name        string `json:"name"`
	Description string `json:"description"`
	Status      string `json:"status"`
	InstanceID  string `json:"instance_id"`
}

type nodeBatchOperateDataJSON struct {
	Creates []nodeCreateDataJSON `json:"objects"`
	Updates []nodeUpdateDataJSON `json:"updates"`
}

type nodeUpdateDataByID struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	Status      string `json:"status"`
	InstanceID  string `json:"instance_id"`
}

type updateNodeParams struct {
	nodeUpdateDataJSON
	ClusterID string `json:"cluster_id" binding:"required"`
}

type nodeListUpdateDataJSON struct {
	Updates []updateNodeParams `json:"updates" binding:"required,dive"`
}

/* node end */

/* namespace start */
type namespaceListForm struct {
	DesireAllData string `form:"desire_all_data"`
	Pagination
}

type createNamespaceDataJSON struct {
	Name           string `binding:"required" json:"name"`
	Creator        string `binding:"required" json:"creator"`
	Description    string `json:"description"`
	EnvType        string `binding:"required" json:"env_type"`
	HasImageSecret bool   `json:"has_image_secret"`
}

type updateNamespaceDataJSON struct {
	Name           string `json:"name"`
	Description    string `json:"description"`
	Status         string `json:"status"`
	EnvType        string `json:"env_type"`
	HasImageSecret *bool  `json:"has_image_secret"`
}

/* namespace end */

/* cluster history start */
type queryClusterHistoryParams struct {
	StartAt time.Time `binding:"required" form:"start_at" time_format:"2006-01-02 15:04:05" time_utc:"1"`
	EndAt   time.Time `binding:"required" form:"end_at" time_format:"2006-01-02 15:04:05" time_utc:"1" validate:"ltecsfield=StartAt"`
	Metric  string    `binding:"required" form:"metric"`
}

/* cluster history end */

// BindForm : binding form with github.com/json-iterator/go
func BindForm(c *gin.Context, obj interface{}) error {
	return c.ShouldBindWith(obj, binding.Form)
}

// BindJSON : binding json with github.com/json-iterator/go
func BindJSON(c *gin.Context, obj interface{}) error {
	return c.ShouldBindWith(obj, binding.JSON)
}

// ValidationError : validate message
func ValidationError(obj interface{}) (err error) {
	result, ok := obj.(validator.ValidationErrors)
	if !ok {
		return fmt.Errorf("%v", obj)
	}
	errStrList := []string{}
	for _, err := range result {
		field := utils.Camel2Snake(err.Field)
		switch err.Tag {
		case "required":
			errStrList = append(errStrList, fmt.Sprintf("param[%s] is null", field))
			continue
		case "max":
			errStrList = append(errStrList, fmt.Sprintf("param[%s] is more than %s", field, err.Param))
			continue
		case "min":
			errStrList = append(errStrList, fmt.Sprintf("param[%s] is less than %s", field, err.Param))
			continue
		}
		errStrList = append(errStrList, fmt.Sprintf("param[%s] is unavailable", field))
	}
	// 返回数据
	return fmt.Errorf("params error, %s", strings.Join(errStrList, ";"))
}

File: bcs-app/backend/apps/metric/serializers.py
# -*- coding: utf-8 -*-
#
# Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community Edition) available.
# Copyright (C) 2017-2019 THL A29 Limited, a Tencent company. All rights reserved.
# Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://opensource.org/licenses/MIT
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
# specific language governing permissions and limitations under the License.
#
import json
import re
import time
from urllib.parse import urlparse

import arrow
from django.utils.translation import ugettext_lazy as _

from backend.apps import constants
from backend.apps.metric.models import Metric
from rest_framework import serializers
from rest_framework.exceptions import ValidationError

NAME_PATTERN = re.compile(r"^[a-z0-9]([-_a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$")


def validate_http_body(body):
    """验证http_body
    """
    try:
        body = json.loads(body)
    except Exception:
        raise ValidationError(_("GET方式，http_body必须是json数据"))

    if not isinstance(body, dict):
        raise ValidationError(_("GET方式，http_body必须是字典类型"))

    for k, v in body.items():
        if isinstance(k, (tuple, list, dict)) or isinstance(v, (tuple, list, dict)):
            raise ValidationError(_("GET方式，http_body键和值必须是数字或者字符串类型"))

        if not k or not v:
            raise ValidationError(_("GET方式，http_body键和值不能为空"))


class UpdateMetricSLZ(serializers.Serializer):
    port = serializers.IntegerField()
    uri = serializers.CharField(min_length=2)
    frequency = serializers.IntegerField()
    http_method = serializers.ChoiceField(choices=Metric.method_choice, default="GET")
    http_headers = serializers.JSONField(required=False, allow_null=True)
    http_body = serializers.CharField(required=False, allow_null=True, allow_blank=True)
    # 新增兼容 prometheus 采集的字段
    metric_type = serializers.CharField(required=False, allow_null=True, allow_blank=True)
    const_labels = serializers.JSONField(required=False, allow_null=True)
    timeout = serializers.IntegerField(required=True)

    def validate_name(self, name):
        count = Metric.objects.filter(project_id=self.context["project_id"], name=name).count()
        if count > 0:
            raise ValidationError(_("name已经存在"))

        if not NAME_PATTERN.match(name):
            raise ValidationError(_("名称由英文字母、下划线或数字组成，且不可以数字开头"))
        return name

    def validate_uri(self, uri):
        parse = urlparse(uri)
        if parse.scheme or parse.netloc:
            raise ValidationError(_("uri不需要填写协议和域名"))
        if uri[0] != "/":
            raise ValidationError(_("uri必须是绝对路径"))
        return uri

    def validate_port(self, port):
        if port < 1 or port > 65535:
            raise ValidationError(_("port必须在1-65535之间"))
        return port

    def validate_http_headers(self, http_headers):
        if not http_headers:
            http_headers = {}
        return json.dumps(http_headers)

    def validate_http_body(self, http_body):
        if not http_body:
            return ""
        else:
            return http_body

    def validate_metric_type(self, metric_type):
        """只能为空或者为'prometheus'
        """
        if not metric_type:
            return ""
        if metric_type in ["prometheus"]:
            return metric_type
        raise ValidationError(_("metric_type 只能为空或者为'prometheus'"))

    def validate_const_labels(self, const_labels):
        if not const_labels:
            const_labels = {}
        return json.dumps(const_labels)

    def validate(self, data):
        if not data:
            raise ValidationError(_("参数不能为空"))

        if data.get("http_body") and data["http_method"] == "GET":
            validate_http_body(data["http_body"])
        return data


class CreateMetricSLZ(UpdateMetricSLZ):
    name = serializers.CharField(max_length=253, min_length=3)

    def validate_name(self, name):
        count = Metric.objects.filter(project_id=self.context["project_id"], name=name).count()
        if count > 0:
            raise ValidationError(_("name已经存在"))

        if not NAME_PATTERN.match(name):
            raise ValidationError(_("名称由英文字母、下划线、中划线或数字组成，且不可以数字开头"))
        return name


class PromMetricSLZBase(serializers.Serializer):
    start_at = serializers.DateTimeField(required=False)
    end_at = serializers.DateTimeField(required=False)

    def validate(self, data):
        now = int(time.time())
        # handle the start_at
        if "start_at" in data:
            data["start_at"] = arrow.get(data["start_at"]).timestamp
        else:
            # default one hour
            data["start_at"] = now - constants.METRICS_DEFAULT_TIMEDELTA
        # handle the end_at
        if "end_at" in data:
            data["end_at"] = arrow.get(data["end_at"]).timestamp
        else:
            data["end_at"] = now
        # start_at must be less than end_at
        if data["end_at"] <= data["start_at"]:
            raise ValidationError(_("param[start_at] must be less than [end_at]"))
        return data


class PromMetricSLZ(PromMetricSLZBase):
    res_id = serializers.CharField(required=True)


class PromPodMetricSLZ(PromMetricSLZBase):
    """Pod数据查询
    """

    res_id_list = serializers.CharField(required=True)

    def validate_res_id_list(self, res_id_list):
        res_id_list = res_id_list.split(",")
        return res_id_list


class PromContainerMetricSLZ(PromMetricSLZBase):
    """容器数据查询
    """

    res_id_list = serializers.CharField(required=False)
    pod_name = serializers.CharField(required=False)

    def validate_res_id_list(self, res_id_list):
        res_id_list = res_id_list.split(",")
        return res_id_list

    def validate(self, data: dict):
        data = super().validate(data)

        if not (data.get("res_id_list") or data.get("pod_name")):
            raise ValidationError(_("res_id_list, pod_name不能同时为空"))

        data.setdefault("pod_name", ".*")
        data.setdefault("res_id_list", [".*"])

        return data


class ServiceMonitorSLZ(serializers.Serializer):
    """ServiceMonitor创建
    """

    name = serializers.CharField()
    cluster_id = serializers.CharField()
    namespace = serializers.CharField()
    service_name = serializers.CharField()
    selector = serializers.JSONField()
    path = serializers.CharField()
    interval = serializers.IntegerField()
    scrape_timeout = serializers.IntegerField()
    sample_limit = serializers.IntegerField()


class ServiceMonitorUpdateSLZ(serializers.Serializer):
    """ServiceMonitor更新
    """

    path = serializers.CharField()
    interval = serializers.IntegerField()
    scrape_timeout = serializers.IntegerField()
    sample_limit = serializers.IntegerField()
    selector = serializers.JSONField()
